### ISSUE TITLE
Add support to new DAPO method

### ISSUE THREAD

**Issue by JohnConnor123** [huggingface/trl#3130]
### Feature request

More info: https://github.com/BytedTsinghua-SIA/DAPO

### Motivation

Good algorithm

### Your contribution

.

**Comment by glowwormX** [comment]
+1

**Comment by 1485840691-eng** [comment]
+1

**Comment by 1485840691-eng** [comment]
Add token_level_loss is straightforward:

        from ..core import masked_mean

        if self.use_token_level_loss:
            loss = masked_mean(per_token_loss, completion_mask)
        else:
            loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()

**Comment by zhing2006** [comment]
+1

**Comment by R4ZZ3** [comment]
+1

**Comment by glowwormX** [comment]
@qgallouedec Thank you for your reply. I have a question about token-level loss. It is the same as [transfomers-34191](https://github.com/huggingface/transformers/issues/34242) When a global batch size contains multiple groups, should token-level loss be summed up among per gpus, per gas, and per microb, and then divided by the number of all tokens? Just do an average once.

**Comment by 1485840691-eng** [comment]
I could help add a simple function as following in grpo_trainer if nobody is working on it.

I wonder if it is fine to use the number of tokens as completion length, which could be obtained  from completion_mask.sum(-1). Given completions is decoded from completion_ids, the actual string length of completions and the number of tokens will be a little different but not much.

Besides, I understand that the reward punishment should be applied on the final reward summing up from all reward function classes. please correct me if it is not the case

rewards = rewards_per_func.sum(dim=1)

> ### Soft Overlong Punishment
> Simply add this reward function. Note that it calculates lengths at string level, but you can also do this at token level by tokenizing completions.
> 
> L_max, L_cache = ..., ...
> 
> def soft_overlong_punishment(completions, **kwargs):
>     rewards = []
>     for completion in completions:
>         completion_length = len(completion)
>         if completion_length <= L_max - L_cache:
>             rewards.append(0)
>         elif L_max - L_cache < completion_length <= L_max:
>             return rewards.append((L_max - L_cache - completion_length) / L_cache)
>         else:
>             return rewards.append(-1)

**Comment by likejazz** [comment]
+1

**Comment by onlyjokers** [comment]
In addition, DAPO's paper mentions that it would be better to disable the KL divergence penalty, but when I try to put beta = 0.0 in GRPOConfig, the loss of the training is always 0

**Comment by michaelpoluektov** [comment]
Do we need to compute `ref_per_token_logps` when KL penalty is disabled?

**Comment by glowwormX** [comment]
> [@qgallouedec](https://github.com/qgallouedec) Thank you for your reply. I have a question about token-level loss. It is the same as [transfomers-34191](https://github.com/huggingface/transformers/issues/34242) When a global batch size contains multiple groups, should token-level loss be summed up among per gpus, per gas, and per microb, and then divided by the number of all tokens? Just do an average once.

@qgallouedec  Is there a plan to fix it?

**Comment by qgallouedec** [comment]
Fixing this problem with the DAPO loss (the one used in trl currently) seems very complicated to me.
On the other hand, it's very likely that we'll soon be using Dr GRPO's loss, which doesn't present this challenge because we don't divide by the number of tokens in the batch, but by the batch size * `max_completion_length`.
Out of curiosity, have you seen very different results because of this problem?

**Comment by glowwormX** [comment]
@qgallouedec 
I haven't experimented yet, and I wonder if the loss in Dr GRPO will cause the loss to be too small. After all, most of the output will not reach max_completion_length. 
And I've got an idea to simply change this to average each time `gradient_accumulation_steps` or per prompt group. That is, all_reduce all lengths at each compute_loss. 
Example code: 
```
denominators = all_reduce(completion_mask.sum()) / world_size
loss = (per_token_loss * completion_mask).sum() / denominators
```
What do you think of that?

**Comment by qgallouedec** [comment]
I understand your point of view. There are three things:
1. It's not a big deal if the loss is small, because you can always play with the learning rate.
2. What you propose corresponds roughly to the `per_loss_batch_norm` in this figure 
![Image](https://github.com/user-attachments/assets/cd1525b9-21e3-4fb7-875c-98b1b6b4f6f7) by Zichen Liu, doesn't it?
If so, as indicated, there's still a batch-level bias.
3. Another problem is the one you pointed out at https://github.com/huggingface/trl/issues/3130#issuecomment-2747142033: when playing with `per_device_train_batch_size` and `gradient_accumulation_steps`, you want the results to be consistent. If you want to divide by the number of tokens in the batch, this makes things really difficult, because you only know the total number of tokens in the batch after the very last time you generate, and by then you've already backpropagated `gradient_accumulation_steps-1` times.

Note that 2. and 3. also applies to the current implementation. The new loss that you suggests mitigates 2. a bit though.

**Comment by glowwormX** [comment]
Thank you for your answer, I now also think that Dr GRPO's loss may be the best option.

**Comment by ArvinZhuang** [comment]
> ### Dynamic Sampling
> Dynamic sampling seems highly inefficient because as training progresses, more samples are generated but ultimately discarded. While it does reduce the number of optimization steps—making results look better when plotted this way, that's the key point here—the real concern is computational efficiency. Over time, dynamic sampling reaches a point where most samples are discarded. A more natural approach would be structuring the dataset in increasing order of difficulty, similar to curriculum learning.

I think ordering data points by difficulty or curriculum learning is not ideal as well. First, it's often hard to know which data point is harder or easier in advance. Second, rollouts of all 1s are not good as well and should be resampled according to DAPO. Maybe dynamic sampling is a more natural solution.

**Comment by zjc664656505** [comment]
I agree. Dynamic Sampling from my perspective is more natural solution. Determining the difficulty level of the data points is hard especially for certain tasks where the label is missing. However DAPO sampling method indeed is inefficient, as they oversample at the beginning and then filter out prompts where all generated outputs are either perfect (accuracy = 1) or completely wrong (accuracy = 0).

**Comment by qgallouedec** [comment]
For overlong filtering, we are experimenting in #3248 and #3229

**Comment by zjc664656505** [comment]
Thanks for the implementation! Wish to contribute as well.
